### ISSUE TITLE
TypeNarrowing on assignment leads to false positive

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#11559]
**Describe the bug**
Hello, I stumbled upon a rather confusing false positive.
It seems assigning to fields shared between multiple types through a union, with a value valid for both (like empty list) the right hand side type is narrowed to one of the field types.

**Code or Screenshots**

```python
class A:
    a: list[int]
class B:
    a: list[str]
def foo(a: A | B):
    a.a = [] # <- empty list is narrowed to list[str]
#   Cannot assign to attribute "a" for class "A"
#   Expression of type "list[str]" cannot be assigned to attribute "a" of class "A"
#     "list[str]" is not assignable to "list[int]"
#       Type parameter "_T@list" is invariant, but "str" is not the same as "int"
#       Consider switching from "list" to "Sequence" which is covariant
#   Pylance: reportAttributeAccessIssue
```

**VS Code extension or command-line**
Found error in pyright 1.1.411 as well as in VSCode pylance

Best regards,
Robin

**Comment by rchiodo** [comment]
Thanks for the report — reproduced, and this is a false positive.

**Root cause:** When the assignment target is a member access whose base is a union (`a: A | B`), pyright has to pick a bidirectional-inference context (an "expected type") for the right-hand side. For a union base it currently just loops the subtypes and keeps one of them (`getDeclaredTypeForExpression` in `typeEvaluator.ts`, which even comments that this case is handled best-effort). Here it picks `B`, so the expected type becomes `list[str]` and the empty list `[]` is inferred as `list[str]`. The assignment is then validated against each subtype separately, and `list[str]` fails against `A`'s `list[int]` — hence the error you see about class `A`.

The tell that it's a bug: pyright arbitrarily chooses `B` to shape the RHS, then reports that the resulting `list[str]` is invalid for `A` — an error it created by its own arbitrary choice. For an empty container the RHS should be inferred per-subtype (`list[int]` for `A`, `list[str]` for `B`), which is exactly why the analogous name-target case produces no error:

```python
def bar(x: list[int] | list[str]):
    x = []  # no error
```

Note that the non-empty cases are correctly flagged, because a concrete invariant list genuinely can't be stored soundly in a field that might be the other type:

```python
def foo(a: A | B):
    a.a = [1]    # correctly errors (not assignable to B's list[str])
    a.a = ["x"]  # correctly errors (not assignable to A's list[int])
```

Working on a fix.
